File: kotlin/canton-sdk/src/main/kotlin/io/github/vsima/canton/CantonClient.kt
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

package io.github.vsima.canton

import com.daml.ledger.api.v2.CommandServiceGrpcKt
import com.daml.ledger.api.v2.CommandServiceOuterClass
import com.daml.ledger.api.v2.StateServiceGrpcKt
import com.daml.ledger.api.v2.StateServiceOuterClass.GetActiveContractsRequest
import com.daml.ledger.api.v2.StateServiceOuterClass.GetLedgerEndRequest
import com.daml.ledger.api.v2.TransactionOuterClass
import com.daml.ledger.api.v2.UpdateServiceGrpcKt
import com.daml.ledger.api.v2.VersionServiceGrpcKt
import com.daml.ledger.api.v2.VersionServiceOuterClass.GetLedgerApiVersionRequest
import io.grpc.CallCredentials
import io.grpc.ManagedChannel
import io.grpc.okhttp.OkHttpChannelBuilder
import io.grpc.stub.AbstractStub
import java.io.Closeable
import java.util.concurrent.TimeUnit
import kotlin.time.Duration
import kotlin.time.TimeSource
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

/**
 * A client for the Canton Ledger API.
 *
 * ```kotlin
 * val client = CantonClient(
 *     CantonClientConfiguration(host = "validator.example.com")
 * )
 * val version = client.ledgerApiVersion()
 * ```
 *
 * The channel-based constructor is useful for tests (in-process transport)
 * and for callers who need full control over the [ManagedChannel].
 */
public class CantonClient(
    private val channel: ManagedChannel,
    callCredentials: CallCredentials? = null,
    private val retryPolicy: RetryPolicy = RetryPolicy.DEFAULT,
    accessTokenProvider: (suspend () -> String)? = null,
) : Closeable {

    public constructor(configuration: CantonClientConfiguration) : this(
        configuration.buildChannel(),
        null,
        configuration.retryPolicy,
        configuration.accessTokenProvider,
    )

    /** Bridges suspending token fetches into gRPC's async credentials API. */
    private val authScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    private val tokenCache: CachingTokenProvider? =
        accessTokenProvider?.let { CachingTokenProvider(fetch = it) }

    private val credentials: CallCredentials? =
        callCredentials ?: tokenCache?.let { BearerTokenCallCredentials(authScope, it::token) }

    private val versionService =
        VersionServiceGrpcKt.VersionServiceCoroutineStub(channel).withAuth(credentials)
    private val commandService =
        CommandServiceGrpcKt.CommandServiceCoroutineStub(channel).withAuth(credentials)
    private val updateService =
        UpdateServiceGrpcKt.UpdateServiceCoroutineStub(channel).withAuth(credentials)
    private val stateService =
        StateServiceGrpcKt.StateServiceCoroutineStub(channel).withAuth(credentials)

    /**
     * Fetches the Ledger API version from the participant.
     *
     * @throws CantonException if the call fails with a gRPC error.
     */
    public suspend fun ledgerApiVersion(): String = withRetries(retryPolicy) {
        mapCantonErrors {
            versionService
                .getLedgerApiVersion(GetLedgerApiVersionRequest.getDefaultInstance())
                .version
        }
    }

    /**
     * Submits [submission] and waits for it to be committed, returning the
     * update id. Retryable failures are retried with the same command id, so
     * the participant deduplicates re-executions.
     *
     * @throws CantonException if the submission ultimately fails.
     */
    public suspend fun submitAndWait(submission: CommandSubmission): String =
        withRetries(retryPolicy) {
            mapCantonErrors {
                commandService.submitAndWait(
                    CommandServiceOuterClass.SubmitAndWaitRequest.newBuilder()
                        .setCommands(submission.toProto())
                        .build()
                ).updateId
            }
        }

    /**
     * Submits [submission], waits for it to be committed, and returns the
     * resulting transaction (flat/ACS-delta shape, filtered to the
     * submitting parties). Retries reuse the same command id.
     *
     * @throws CantonException if the submission ultimately fails.
     */
    public suspend fun submitAndWaitForTransaction(
        submission: CommandSubmission,
    ): TransactionOuterClass.Transaction =
        withRetries(retryPolicy) {
            mapCantonErrors {
                commandService.submitAndWaitForTransaction(
                    CommandServiceOuterClass.SubmitAndWaitForTransactionRequest.newBuilder()
                        .setCommands(submission.toProto())
                        .build()
                ).transaction
            }
        }

    /**
     * The participant's current ledger end offset — the natural
     * [UpdateSubscription.beginExclusive] for a fresh subscription.
     *
     * @throws CantonException if the call fails with a gRPC error.
     */
    public suspend fun ledgerEnd(): Long = withRetries(retryPolicy) {
        mapCantonErrors {
            stateService.getLedgerEnd(GetLedgerEndRequest.getDefaultInstance()).offset
        }
    }

    /**
     * The active contract set visible to [parties] at [activeAtOffset]
     * (all templates). Retryable failures restart the snapshot from scratch,
     * so the result is never partial.
     *
     * @throws CantonException if the call ultimately fails.
     */
    public suspend fun activeContracts(
        parties: List<String>,
        activeAtOffset: Long,
        verbose: Boolean = true,
    ): List<ActiveContract> = withRetries(retryPolicy) {
        mapCantonErrors {
            val request = GetActiveContractsRequest.newBuilder()
                .setActiveAtOffset(activeAtOffset)
                .setEventFormat(wildcardEventFormat(parties, verbose))
                .build()
            val contracts = mutableListOf<ActiveContract>()
            stateService.getActiveContracts(request).collect { response ->
                if (response.hasActiveContract()) {
                    val entry = response.activeContract
                    contracts += ActiveContract(
                        createdEvent = entry.createdEvent,
                        synchronizerId = entry.synchronizerId,
                        reassignmentCounter = entry.reassignmentCounter,
                    )
                }
            }
            contracts
        }
    }

    /**
     * The active contract set at the current ledger end — the starting point
     * for a full state sync. Apply [ActiveContractsSnapshot.contracts], then
     * consume [updates] from [ActiveContractsSnapshot.offset]:
     *
     * ```kotlin
     * val snapshot = client.activeContractsSnapshot(listOf(party))
     * // apply snapshot.contracts ...
     * client.updates(UpdateSubscription(listOf(party), beginExclusive = snapshot.offset))
     *     .collect { /* deltas, gap-free */ }
     * ```
     */
    public suspend fun activeContractsSnapshot(
        parties: List<String>,
        verbose: Boolean = true,
    ): ActiveContractsSnapshot {
        val offset = ledgerEnd()
        return ActiveContractsSnapshot(offset, activeContracts(parties, offset, verbose))
    }

    /**
     * Streams ledger updates for [subscription], transparently reconnecting
     * on retryable failures and resuming from the offset of the last
     * received update — consumers see one uninterrupted, gap-free stream.
     * The retry budget resets only once a connection has proven healthy:
     * it delivered at least one update and stayed alive for
     * [RetryPolicy.streamHealthyWindow]. Shorter-lived connections keep
     * escalating the backoff until [RetryPolicy.maxAttempts] is exhausted.
     *
     * The flow completes normally when the server ends the stream (only for
     * subscriptions with [UpdateSubscription.endInclusive] set) and throws
     * [CantonException] on non-retryable failures.
     */
    public fun updates(subscription: UpdateSubscription): Flow<LedgerUpdate> = flow {
        var cursor = subscription.beginExclusive
        var attempt = 1
        var authRecovered = false
        while (true) {
            val connectedAt = TimeSource.Monotonic.markNow()
            var progressed = false
            try {
                updateService.getUpdates(subscription.toRequest(cursor)).collect { response ->
                    val update = LedgerUpdate.from(response) ?: return@collect
                    cursor = update.offset
                    progressed = true
                    emit(update)
                }
                return@flow // server completed the stream (finite subscription)
            } catch (t: Throwable) {
                val error = CantonError.from(t) ?: throw t
                // Only a connection that delivered an update and outlived the
                // healthy window earns a fresh retry budget; a stream that
                // reconnects and dies young keeps escalating instead.
                if (progressed && connectedAt.elapsedNow() >= retryPolicy.streamHealthyWindow) {
                    attempt = 1
                }
                // Auth recovery re-arms on lifetime alone: a connection the
                // server accepted and held past the healthy window proves
                // the last refresh worked even if the ledger was idle —
                // its token simply aged out again.
                if (connectedAt.elapsedNow() >= retryPolicy.streamHealthyWindow) {
                    authRecovered = false
                }
                // The participant terminates a stream whose access token
                // expired mid-flight (observed as PERMISSION_DENIED, no
                // RetryInfo). That is recoverable exactly when the provider
                // can mint a token that differs from the one the failed
                // connection used — its fetch has populated the cache by
                // now, so cached() is that token. Reconnect once with the
                // fresh one; a second auth failure without an intervening
                // healthy connection is a real authorization problem and
                // propagates.
                if (error.isAuthFailure && tokenCache != null && !authRecovered &&
                    tokenCache.refreshIfChanged(tokenCache.cached())
                ) {
                    authRecovered = true
                    continue
                }
                if (!error.retryable || attempt >= retryPolicy.maxAttempts) {
                    throw CantonException(error, t)
                }
                delay(maxOf(retryPolicy.backoffFor(attempt), error.retryDelay ?: Duration.ZERO))
                attempt++
            }
        }
    }

    override fun close() {
        authScope.cancel()
        channel.shutdown()
        if (!channel.awaitTermination(5, TimeUnit.SECONDS)) {
            channel.shutdownNow()
        }
    }
}

private fun <S : AbstractStub<S>> S.withAuth(credentials: CallCredentials?): S =
    if (credentials != null) withCallCredentials(credentials) else this

private fun CantonClientConfiguration.buildChannel(): ManagedChannel =
    OkHttpChannelBuilder.forAddress(host, port)
        .apply { if (useTls) useTransportSecurity() else usePlaintext() }
        .build()

File: swift/Sources/CantonKit/CantonClient.swift
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

import CantonLedgerAPI
import GRPCCore
import GRPCNIOTransportHTTP2

/// A client for the Canton Ledger API.
///
/// `CantonClient` owns connection configuration and exposes scoped access to
/// the generated Ledger API service clients:
///
/// ```swift
/// let client = CantonClient(
///     configuration: .init(host: "validator.example.com", useTLS: true)
/// )
/// let version = try await client.ledgerApiVersion()
/// ```
public struct CantonClient: Sendable {
    #if canImport(Network)
    /// Network.framework-backed transport: better cellular/Wi-Fi handoff and
    /// VPN behavior on Apple platforms.
    public typealias Transport = HTTP2ClientTransport.TransportServices
    #else
    public typealias Transport = HTTP2ClientTransport.Posix
    #endif

    public let configuration: CantonClientConfiguration

    public init(configuration: CantonClientConfiguration) {
        self.configuration = configuration
    }

    /// Ledger API service clients bound to an open connection.
    public struct Services: Sendable {
        /// The underlying gRPC client; wrap any generated service client
        /// around this for APIs not yet surfaced by CantonKit.
        public let grpc: GRPCClient<Transport>

        public var version: Com_Daml_Ledger_Api_V2_VersionService.Client<Transport> {
            .init(wrapping: grpc)
        }

        public var command: Com_Daml_Ledger_Api_V2_CommandService.Client<Transport> {
            .init(wrapping: grpc)
        }

        public var state: Com_Daml_Ledger_Api_V2_StateService.Client<Transport> {
            .init(wrapping: grpc)
        }

        public var update: Com_Daml_Ledger_Api_V2_UpdateService.Client<Transport> {
            .init(wrapping: grpc)
        }
    }

    /// Opens a connection, runs `body` against the Ledger API services, and
    /// shuts the connection down when `body` returns.
    public func withServices<Result: Sendable>(
        _ body: @Sendable (Services) async throws -> Result
    ) async throws -> Result {
        var interceptors: [any ClientInterceptor] = []
        if let tokenProvider = configuration.accessTokenProvider {
            interceptors.append(BearerTokenInterceptor(tokenProvider: tokenProvider))
        }
        #if canImport(Network)
        let transport: Transport = try .http2NIOTS(
            target: .dns(host: configuration.host, port: configuration.port),
            transportSecurity: configuration.useTLS ? .tls : .plaintext
        )
        #else
        let transport: Transport = try .http2NIOPosix(
            target: .dns(host: configuration.host, port: configuration.port),
            transportSecurity: configuration.useTLS ? .tls : .plaintext
        )
        #endif
        return try await withGRPCClient(
            transport: transport,
            interceptors: interceptors
        ) { grpc in
            try await body(Services(grpc: grpc))
        }
    }

    /// Convenience: fetches the Ledger API version from the participant.
    ///
    /// - Throws: ``CantonError`` if the call fails with a gRPC error.
    public func ledgerApiVersion() async throws -> String {
        try await withRetries(configuration.retryPolicy) {
            try await mapCantonErrors {
                try await withServices { services in
                    try await services.version.getLedgerApiVersion(.init()).version
                }
            }
        }
    }

    /// Submits `submission` and waits for it to be committed, returning the
    /// update id. Retryable failures are retried with the same command id,
    /// so the participant deduplicates re-executions.
    ///
    /// - Throws: ``CantonError`` if the submission ultimately fails.
    public func submitAndWait(_ submission: CommandSubmission) async throws -> String {
        try await withRetries(configuration.retryPolicy) {
            try await mapCantonErrors {
                try await withServices { services in
                    var request = Com_Daml_Ledger_Api_V2_SubmitAndWaitRequest()
                    request.commands = submission.proto
                    return try await services.command.submitAndWait(request).updateID
                }
            }
        }
    }

    /// The participant's current ledger end offset — the natural
    /// ``UpdateSubscription/beginExclusive`` for a fresh subscription.
    ///
    /// - Throws: ``CantonError`` if the call fails with a gRPC error.
    public func ledgerEnd() async throws -> Int64 {
        try await withRetries(configuration.retryPolicy) {
            try await mapCantonErrors {
                try await withServices { services in
                    try await services.state.getLedgerEnd(.init()).offset
                }
            }
        }
    }

    /// The active contract set visible to `parties` at `activeAtOffset`
    /// (all templates). Retryable failures restart the snapshot from
    /// scratch, so the result is never partial.
    ///
    /// - Throws: ``CantonError`` if the call ultimately fails.
    public func activeContracts(
        parties: [String],
        activeAtOffset: Int64,
        verbose: Bool = true
    ) async throws -> [ActiveContract] {
        try await withRetries(configuration.retryPolicy) {
            try await mapCantonErrors {
                try await withServices { services in
                    var request = Com_Daml_Ledger_Api_V2_GetActiveContractsRequest()
                    request.activeAtOffset = activeAtOffset
                    request.eventFormat = wildcardEventFormat(parties: parties, verbose: verbose)
                    return try await services.state.getActiveContracts(request) { response in
                        var contracts: [ActiveContract] = []
                        for try await message in response.messages {
                            guard case .activeContract(let entry)? = message.contractEntry else { continue }
                            contracts.append(
                                ActiveContract(
                                    createdEvent: entry.createdEvent,
                                    synchronizerId: entry.synchronizerID,
                                    reassignmentCounter: entry.reassignmentCounter
                                )
                            )
                        }
                        return contracts
                    }
                }
            }
        }
    }

    /// The active contract set at the current ledger end — the starting
    /// point for a full state sync. Apply the snapshot's contracts, then
    /// consume ``updates(_:)`` from the snapshot's offset for gap-free deltas.
    public func activeContractsSnapshot(
        parties: [String],
        verbose: Bool = true
    ) async throws -> ActiveContractsSnapshot {
        let offset = try await ledgerEnd()
        return ActiveContractsSnapshot(
            offset: offset,
            contracts: try await activeContracts(parties: parties, activeAtOffset: offset, verbose: verbose)
        )
    }

    /// Streams ledger updates for `subscription`, transparently reconnecting
    /// on retryable failures and resuming from the offset of the last
    /// received update — consumers see one uninterrupted, gap-free stream.
    /// The retry budget resets only once a connection has proven healthy:
    /// it delivered at least one update and stayed alive for
    /// ``RetryPolicy/streamHealthyWindow``. Shorter-lived connections keep
    /// escalating the backoff until ``RetryPolicy/maxAttempts`` is exhausted.
    ///
    /// The stream finishes normally when the server ends it (only for
    /// subscriptions with ``UpdateSubscription/endInclusive`` set) and throws
    /// ``CantonError`` on non-retryable failures.
    public func updates(_ subscription: UpdateSubscription) -> AsyncThrowingStream<LedgerUpdate, any Error> {
        let policy = configuration.retryPolicy

        func retryDelay(for error: any Error, attempt: Int) -> Duration? {
            guard let canton = CantonError(error), canton.isRetryable, attempt < policy.maxAttempts else {
                return nil
            }
            var delay = policy.backoff(forAttempt: attempt)
            if let suggested = canton.retryDelay, suggested > delay {
                delay = suggested
            }
            return delay
        }

        return AsyncThrowingStream { continuation in
            let task = Task {
                let clock = ContinuousClock()
                var cursor = subscription.beginExclusive
                var attempt = 1
                while !Task.isCancelled {
                    let begin = cursor
                    let connectedAt = clock.now
                    do {
                        let (last, progressed, streamError) = try await self.withServices { services in
                            try await services.update.getUpdates(subscription.request(from: begin)) { response in
                                var last = begin
                                var progressed = false
                                do {
                                    for try await message in response.messages {
                                        guard let update = LedgerUpdate(message) else { continue }
                                        last = update.offset
                                        progressed = true
                                        continuation.yield(update)
                                    }
                                } catch {
                                    return (last, progressed, error as (any Error)?)
                                }
                                return (last, progressed, nil)
                            }
                        }
                        cursor = last
                        // Only a connection that delivered an update and
                        // outlived the healthy window earns a fresh retry
                        // budget; a stream that reconnects and dies young
                        // keeps escalating instead.
                        if progressed, clock.now - connectedAt >= policy.streamHealthyWindow {
                            attempt = 1
                        }
                        guard let streamError else {
                            continuation.finish() // server completed (finite subscription)
                            return
                        }
                        guard !(streamError is CancellationError),
                              let delay = retryDelay(for: streamError, attempt: attempt)
                        else {
                            continuation.finish(throwing: CantonError(streamError) ?? streamError)
                            return
                        }
                        try await Task.sleep(for: delay)
                        attempt += 1
                    } catch is CancellationError {
                        break
                    } catch {
                        guard let delay = retryDelay(for: error, attempt: attempt) else {
                            continuation.finish(throwing: CantonError(error) ?? error)
                            return
                        }
                        do {
                            try await Task.sleep(for: delay)
                        } catch {
                            break
                        }
                        attempt += 1
                    }
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// Submits `submission`, waits for it to be committed, and returns the
    /// resulting transaction (flat/ACS-delta shape, filtered to the
    /// submitting parties). Retries reuse the same command id.
    ///
    /// - Throws: ``CantonError`` if the submission ultimately fails.
    public func submitAndWaitForTransaction(
        _ submission: CommandSubmission
    ) async throws -> Com_Daml_Ledger_Api_V2_Transaction {
        try await withRetries(configuration.retryPolicy) {
            try await mapCantonErrors {
                try await withServices { services in
                    var request = Com_Daml_Ledger_Api_V2_SubmitAndWaitForTransactionRequest()
                    request.commands = submission.proto
                    return try await services.command.submitAndWaitForTransaction(request).transaction
                }
            }
        }
    }
}

/// Injects `authorization: Bearer <token>` into every request, fetching the
/// token from the configured provider.
struct BearerTokenInterceptor: ClientInterceptor {
    let tokenProvider: @Sendable () async throws -> String

    func intercept<Input: Sendable, Output: Sendable>(
        request: StreamingClientRequest<Input>,
        context: ClientContext,
        next: (StreamingClientRequest<Input>, ClientContext) async throws -> StreamingClientResponse<Output>
    ) async throws -> StreamingClientResponse<Output> {
        var request = request
        let token = try await tokenProvider()
        request.metadata.replaceOrAddString("Bearer \(token)", forKey: "authorization")
        return try await next(request, context)
    }
}
